### DISCUSSION TITLE
About Visible Spectrum Volume Boundaries.

### DISCUSSION THREAD

**Discussion by thomasmansencal** [colour-science/colour#708] [General]
I'm starting a thread in regard to https://stackoverflow.com/a/48396021/931625 as we will hit comments limit there.

I'm assuming @nschloe is author of the question and the snippet being discussed is as follows:

``` python
import colour
import numpy as np


def generate_square_waves(samples):
    square_waves = []
    square_waves_basis = np.tril(np.ones((samples, samples)))[0:-1, :]
    for i in range(samples):
        square_waves.append(np.roll(square_waves_basis, i))
    return np.vstack((np.zeros(samples), np.vstack(square_waves),
                    np.ones(samples)))


def XYZ_outer_surface(samples):
    XYZ = []
    wavelengths = np.linspace(colour.DEFAULT_SPECTRAL_SHAPE.start,
                            colour.DEFAULT_SPECTRAL_SHAPE.end, samples)

    for wave in generate_square_waves(samples):
        spd = colour.SpectralPowerDistribution(
            wave, wavelengths, interpolator=colour.LinearInterpolator).align(
                colour.DEFAULT_SPECTRAL_SHAPE)
        XYZ.append(colour.spectral_to_XYZ(spd))

    return np.array(XYZ).reshape(len(XYZ), -1, 3)


# 43 is picked as number of samples to have integer wavelengths.
colour.write_image(XYZ_outer_surface(43), 'CIE_XYZ_outer_surface.exr')
```

The output image is here: https://drive.google.com/file/d/1GScSAMyyljDQGvZqUWluix9n9fy66NNs/view?usp=sharing

**Comment by nschloe** [comment]
What I'm getting here is
```
Traceback (most recent call last):
  File "b.py", line 29, in <module>
    colour.write_image(XYZ_outer_surface(43), 'CIE_XYZ_outer_surface.exr')
  File "b.py", line 21, in XYZ_outer_surface
    wave, wavelengths, interpolator=colour.LinearInterpolator).align(
TypeError: __init__() got an unexpected keyword argument 'interpolator'
```
This is with
```
$ python3 -c "import colour; print(colour.__version__)"
0.3.10
```
I can get it do something with
```python
import colour
import numpy as np


def generate_square_waves(samples):
    square_waves = []
    square_waves_basis = np.tril(np.ones((samples, samples)))[0:-1, :]
    for i in range(samples):
        square_waves.append(np.roll(square_waves_basis, i))
    return np.vstack((
            np.zeros(samples), np.vstack(square_waves), np.ones(samples)
            ))


def XYZ_outer_surface(samples):
    XYZ = []
    wavelengths = np.linspace(
            colour.DEFAULT_SPECTRAL_SHAPE.start,
            colour.DEFAULT_SPECTRAL_SHAPE.end, samples
            )

    for wave in generate_square_waves(samples):
        spd = colour.SpectralPowerDistribution(
            'custom',
            {wl: value for wl, value in zip(wavelengths, wave)}
            # interpolator=colour.LinearInterpolator
            ).align(colour.DEFAULT_SPECTRAL_SHAPE)
        XYZ.append(colour.spectral_to_XYZ(spd))

    return np.array(XYZ).reshape(len(XYZ), -1, 3)


# 43 is picked as number of samples to have integer wavelengths.
colour.write_image(XYZ_outer_surface(43), 'CIE_XYZ_outer_surface.exr')
```

**Comment by thomasmansencal** [comment]
This is my fault, I should have mentioned that the code is using latest [develop](https://github.com/colour-science/colour/tree/develop) branch. We have changed the whole spectral computations backend and there are some backward incompatible changes.

I could adjust the code for v0.3.10 if convenient for you though?

**Comment by nschloe** [comment]
Let me just grab develop and try it out.

**Comment by nschloe** [comment]
Alright, same result: exrdisplay shows
![w](https://user-images.githubusercontent.com/181628/35829823-6d8878d8-0ac4-11e8-9042-fc2379b8ecb8.png)
Perhaps I should use another viewer? I'm not at all familiar with exr files.

**Comment by thomasmansencal** [comment]
Oh right, this is the expected output, it is essentially a one pixel width image storing all the XYZ coordinates. I have used that because I can load it in [The Foundry Nuke](https://www.foundry.com/products/nuke) easily and it was more convenient to illustrate the SO thread than pushing data in MPL or Plot.ly.

Here is a Plot.ly example: https://plot.ly/~KelSolaar/91 and the generating code:

``` python
import plotly.plotly as py
import plotly.graph_objs as go

# From above "XYZ_outer_surface" definition.
XYZ = XYZ_outer_surface(43)

X, Y, Z = colour.utilities.tsplit(XYZ)
trace = go.Scatter3d(x=X, y=Y, z=Z)

data = [trace]
layout = go.Layout(margin=dict(l=0, r=0, b=0, t=0))
figure = go.Figure(data=data, layout=layout)
py.iplot(figure, filename='Visible Spectrum Volume Boundaries')

```

**Comment by nschloe** [comment]
Alright, I think I get it now. Here's an image of a small test code I wrote:
![d](https://user-images.githubusercontent.com/181628/35830374-516b36ac-0ac6-11e8-8aaf-5c192045f946.png)
One generates "block spectra" which are 0 everywhere except in a certain range of wavelengths, and computed the XYZ coordinates of that. The value of the block is simply set _some_ constant, e.g., 1; it doesn't really matter for the shape.

With reference to the original question on SO, I'm wondering now what the significance of that shape is at all. I mean, the fact that one sets the limit to 1 doesn't really constitute something physical, right? One could easily just crank up some wavelengths here and there, and bam, the blob expands.

**Comment by thomasmansencal** [comment]
> I mean, the fact that one sets the limit to 1 doesn't really constitute something physical, right?

As discussed on SO, it is a normalisation process, so you could scale the volume uniformly ideally.

> I could easily just crank up some wavelengths here and there, and bam, the blob expands.

Yes although it would not yield a very useful volume representation! :)

**Comment by nschloe** [comment]
Yes, I guess you just scale the thing such that the full spectrum maps to [100,100,100]. Fair enough.

Let me say thanks _very_ much for the discussion! This really enhanced my understanding of color theory.

**Comment by thomasmansencal** [comment]
You are very much welcome!

**Comment by thomasmansencal** [comment]
Oh and before I forget if you are happy about the SO answer, would you mind accepting it please? 

Cheers,

Thomas

**Comment by nschloe** [comment]
> Oh and before I forget if you are happy about the SO answer, would you mind accepting it please?

Yeah, sure. Before that, it probably needs some cleaning up, and give the answer, too. :) For example:

> The upper boundary is the result of CIE Y, the central Colour Matching Function of the CIE 1931 2 Degree Standard Observer being normalised to 1.

I don't think this is true, is it? The maximum of CIE Y is 1 alright, but even if the functions were scaled in any other way, the shape would come up. It is rather a consequence of the fact that one restricts oneself to two values in the spectrum, 0 and alpha (where alpha is arbitrarily chosen), right?

Also, the answer should probably mention that "gamut of visible colors" is a bit of a misnomer. It's the gamut of colors generated by spectra with a certain maximum value.

All this is more or less mentioned in the section _Outer Surface Generation_ of the answer. This part should probably be moved to the beginning of the reply.

**Comment by nschloe** [comment]
Ah, I think now I understood something else: The visible gamut is the visible gamut _under a particular illuminant!_ Taking an "arbitrary" reference spectrum of all ones, corresponding to a perfectly white object, the integral
```
∫_360nm^830nm observer(λ) illumination(λ) dλ
```
will give you the "whitest" color possible – under that illuminant. You're not just feeding square spectra into the observer, but chunks of the illuminant!

Setting the illuminant to 1, as we have to so far, is essentially equivalent to picking the rarely used E illuminant.

So, yeah, I guess it _does_ make sense after all to speak about the gamut of visible colors (under a particular illuminant).

**Comment by thomasmansencal** [comment]
> I don't think this is true, is it? The maximum of CIE Y is 1 alright, but even if the functions were scaled in any other way, the shape would come up. It is rather a consequence of the fact that one restricts oneself to two values in the spectrum, 0 and alpha (where alpha is arbitrarily chosen), right?

Yes, I need to reformulate that part to account for the whole surface not just the upper limit, I originally did not understand what you were originally looking for.

**Comment by thomasmansencal** [comment]
> The visible gamut is the visible gamut under a particular illuminant!

Not really under a particular illuminant but under the Equal Energy illuminant which have constant radiant power.

The [CIE definition](http://eilv.cie.co.at/term/236) where every word is important is actually as follows:

**The colour matching functions are the tristimulus values of monochromatic stimuli of equal radiant power.**

It is the first sentence of our dedicated (not-up-to-date) notebook here: https://github.com/colour-science/colour-notebooks/blob/master/notebooks/colorimetry/cmfs.ipynb

There is an important but critical subtlety when you compute the tristimulus values XYZ of a sample, the **illuminant** part is actually the illuminant S under which you are viewing the sample, it is independent of the CMFS:

![image](https://user-images.githubusercontent.com/99779/35880031-a38f0302-0be1-11e8-89c6-e3e1946ba9ff.png)

Notice that there is no parameterisation _to include an hypothetical illuminant under which the CMFS would have been measured_, they are _pure_. You can replace them with camera sensor sensitivities and still not have to include any illuminant during the measurement conditions for the sensitivities.

This is the reason I was saying "crank(ing) up some wavelengths here and there" would not be useful because you would be artificially increasing HVS sensitivity at some particular wavelengths.

This page has some really interesting further reading on how the CMFS were derived through colour matching experiments by Wright & Guild following Maxwell work: http://www.handprint.com/HP/WCL/color6.html

**Comment by nschloe** [comment]
Thanks again for the replies. I think we're on the same page now.

>> The visible gamut is the visible gamut under a particular illuminant!
>
>Not really under a particular illuminant but under the Equal Energy illuminant which have constant radiant power.

Alright. What I meant to say is: It doesn't make sense to speak of the visible gamut _per se_, but only with an illuminant; equal energy, D65, pick your poison. That illuminant then shines upon the object to consider, and it reflects the wavelengths more or less well (β(λ)).

> The colour matching functions are the tristimulus values of monochromatic stimuli of equal radiant power.

"Equal radiant power" actually means I can pick and, because at the end of the day they are all normalized by `Y_W=100`, right?

> Notice that there is no parameterisation to include an hypothetical illuminant under which the CMFS would have been measured, they are pure.

I don't know what that wants to say.

> This page has some really interesting further reading on how the CMFS were derived through colour matching experiments by Wright & Guild following Maxwell work: http://www.handprint.com/HP/WCL/color6.html

Looks super interesting, thanks for the hint!

**Comment by thomasmansencal** [comment]
> It doesn't make sense to speak of the visible gamut per se

We are probably entering the meta-discussion stage here. The way I see it is as follows, the HVS is a sensory system whose purpose is to measure something thus there is an envelope to the extent of what it is capable of measuring. The visible gamut represents the envelope for an event that would stimulate the sensor fully. It is obviously more complex because the envelope is relative and we don't account for saturation, etc...

> I don't know what that wants to say.

What I was trying to say is that when you use the CMFS, you don't care about the conditions under which they were assembled and put together. They are pure in that they are self-contained like a blackbox and you don't have any parameter to change the way they measure radiant energy. A parallel would be a thermometer, in order to use one, you don't need knowledge of how it was assembled and you don't have hooks to tweak the way it measure temperature, it simply measure it.

Does it make sense?

**Comment by nschloe** [comment]
> Does it make sense?

Sure, feel free to close this one. For your extraordinary effort, I've awarded a 100 points bounty on stackoverflow. Thank you again!

**Comment by thomasmansencal** [comment]
Thanks @nschloe, this is much appreciated!

**Comment by yuhao** [comment]
Hi came across this when searching for how the visual gamut boundary is calculated. Good discussion; thanks. I have a few points to make though.

It does seems arbitrary that the visual gamut is sampled at a few squared waves. In theory, we should be exhausting all physically-possible spectra and calculate the corresponding LMS/XYZ values and then plot them. That's obviously not possible and so sampling is necessary. But I am not sure I follow @KelSolaar's point that "crank(ing) up some wavelengths here and there would not be useful because you would be artificially increasing HVS sensitivity at some particular wavelengths." I am not sure how that's related to HVS sensitivity at all. For any given arbitrary spectrum, I can use the cone sensitivities to calculate the LMS responses for that light, which corresponds to a point in the LMS space and is a point that's inside the visible gamut. Why is using an arbitrary light artificially increasing HVS sensitivity?

Also, I am not sure I follow the following discussion:

> >The visible gamut is the visible gamut under a particular illuminant!
>
>Not really under a particular illuminant but under the Equal Energy illuminant which have constant radiant power.
>
>The CIE definition where every word is important is actually as follows:
>
>The colour matching functions are the tristimulus values of monochromatic stimuli of equal radiant power.

Visible gamut has nothing to do with illuminant doesn't it? Visible gamut is just all the colors that HVS can see.

I am not sure why the definition of CMF has anything to do with visible gamut, either. The triplets at each wavelength in the CMFs tells us the responses stimulated from different spectral (monochromatic) lights that have the same energy/power/irradiance/radiance. So the tristimulus values at each wavelength in the CMFs are responses of unit-power monochromatic lights, not equal-energy illuminant. This interactive tutorial explains what CMFs really are: https://www.cs.rochester.edu/courses/572/colorvis/cone2cmf.html.

> This page has some really interesting further reading on how the CMFS were derived through colour matching experiments by Wright & Guild following Maxwell work: http://www.handprint.com/HP/WCL/color6.html

That's a great piece of work. I'd also recommend this article: http://yuhaozhu.com/blog/cmf.html, which is based on two prior publications, both of which meticulously document the exact steps taken to convert Wright & Guild's experimental data to derive the CIE 1931 RGB CMFs.

* [Broadbent 2004a] A Critical Review of the Development of the CIE1931 RGB Color-Matching Functions. Color Research & Application 29(4):267–272. August 2004.
* The Wright – Guild Experiments and the Development of the CIE 1931 RGB and XYZ Color Spaces: https://philservice.typepad.com/Wright-Guild_and_CIE_RGB_and_XYZ.pages.pdf.

**Reply by thomasmansencal** [reply]
Hi @yuhao,

>  But I am not sure I follow @KelSolaar's point that "crank(ing) up some wavelengths here and there would not be useful because you would be artificially increasing HVS sensitivity at some particular wavelengths."

Let's assume that we compute the visible spectrum volume boundaries not with the Equal Energy Illuminant but a laser light with a peak at 555nm. The boundary will have an extremely distorted shape which is akin to artificially increase the sensitivity for that particular wavelength or conversely decrease it for others.

> Visible gamut has nothing to do with illuminant doesn't it? Visible gamut is just all the colors that HVS can see.

Yes but nothing says either that you cannot compute the boundaries for a particular illuminant. It might be contextually useful.

Cheers,

Thomas

**Reply by yuhao** [reply]
Hi @KelSolaar,

Thanks for the response. So my understanding is that HVS gamut is unlimited and so it doesn't quite have a boundary/shape per se. So any visualization of its "boundary" must be scenario-dependent and, as you rightfully pointed out, be contextually useful. I was curious in what context is it useful to visualize the boundary by using a set of square spectra? Those spectra stimulate the cones across a **continuous** band --- what's the significance of a continuous band? Why wouldn't it be useful to use, say, a spectrum where the energy is 1 between say 500nm and 600nm and between 650nm and 700nm and 0 everywhere else?

**Reply by thomasmansencal** [reply]
> So my understanding is that HVS gamut is unlimited and so it doesn't quite have a boundary/shape per se

Well, it is a bit rhetorical but there are certainly absolute limits to what the HVS is sensitivity to, e.g. cones detection threshold to the point where they saturate and bleach. Besides that, you can give a constrain to the input signal, e.g. taking slices of the equal energy illuminant and produce the boundary for those stimuli, in this case, producing the *Rösch-MacAdam* colour solid.

> Those spectra stimulate the cones across a continuous band --- what's the significance of a continuous band?

The square pulses allow to conveniently sample the entire surface, depending how you build/order them you can directly generate the orange surface here:

![image](https://user-images.githubusercontent.com/99779/150719838-ba9e45e9-0e1e-4536-95de-dcda2058d282.png)

or

![image](https://user-images.githubusercontent.com/99779/150718815-d496e1ec-d811-4f19-b909-9e3f53d71840.png)

Which has some similarity with:

![image](https://user-images.githubusercontent.com/99779/150718861-3ddf1f7c-b9c0-42e3-b929-92f76abce921.png)

And share traits with _Hellwig and Fairchild (2020)_ in **Gaussian Spectra to Derive a Hue-linear Color Space**
